src/lavinmq/auth/authenticators/oauth.cr

```diff
@@ -0,0 +1,317 @@
+require "../authenticator"
+require "../users/oauth_user"
+require "../../config"
+require "../jwt"
+require "openssl"
+require "http/client"
+require "../lib_crypto_ext"
+
+module LavinMQ
+  module Auth
+    class OAuthAuthenticator < Authenticator
+      Log = LavinMQ::Log.for "oauth2"
+
+      @cached_keys : Hash(String, String)?
+      @cache_expires_at : Time?
+      @cache_mutex : Mutex = Mutex.new
+
+      def initialize(@users : Auth::UserStore, @config = Config.instance)
+      end
+
+      def authenticate(username : String, password : Bytes) : OAuthUser?
+        token = fetch_and_verify_jwks(String.new(password))
+        username, tags, permissions, expires_at = parse_jwt_payload(token.payload)
+
+        # Validate token expiration
+        expiration_time = Time.unix(expires_at)
+        if expiration_time <= Time.utc
+          Log.warn { "OAuth2 authentication failed: Token has already expired" }
+          return nil
+        end
+
+        Log.info { "OAuth2 user authenticated: #{username}" }
+        OAuthUser.new(username, tags, permissions, expiration_time)
+      rescue ex : JWT::DecodeError
+        Log.warn { "OAuth2 authentication failed: Could not decode token - #{ex.message}" }
+        nil
+      rescue ex : JWT::VerificationError
+        Log.warn { "OAuth2 authentication failed: Token verification failed - #{ex.message}" }
+        nil
+      rescue ex : Exception
+        Log.error { "OAuth2 authentication failed: #{ex.message}" }
```

Comment:
Since this is an unexpected error we should include the exception
```suggestion
        Log.error(exception: ex) { "OAuth2 authentication failed for user \"#{username}\": #{ex.message}" }
```